ds/reactive/sorted_set_impl.go

```diff
@@ -0,0 +1,252 @@
+				defer s.mutex.Unlock()
+			}
+
+			listElement.weight = newWeight
+
+			s.updatePosition(listElement)
+		}, true)
+	}
+}
+
+// deleteSorted deletes the given element from the sortedElements slice.
+func (s *sortedSet[ElementType, WeightType]) deleteSorted(element ElementType) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	if deletedElement, deleted := s.elements.DeleteAndReturn(element); deleted {
+		// unsubscribe from weight updates
+		deletedElement.unsubscribeFromWeightUpdates()
+
+		// shift all elements to the right of the deleted element one position to the left
+		for i := deletedElement.index; i < len(s.sortedElements)-1; i++ {
+			s.sortedElements[i] = s.sortedElements[i+1]
+			s.sortedElements[i].index--
+		}
+
+		// prevent memory leak and shrink slice
+		s.sortedElements[len(s.sortedElements)-1] = nil
+		s.sortedElements = s.sortedElements[:len(s.sortedElements)-1]
+
+		// update heaviest and lightest element
+		if deletedElement.index == 0 {
+			if len(s.sortedElements) > 0 {
+				s.heaviestElement.Set(s.sortedElements[0].element)
+			} else {
+				s.heaviestElement.Set(*new(ElementType))
+			}
+		}
+		if deletedElement.index == len(s.sortedElements) {
+			if len(s.sortedElements) > 0 {
+				s.lightestElement.Set(s.sortedElements[len(s.sortedElements)-1].element)
+			} else {
+				s.lightestElement.Set(*new(ElementType))
+			}
+		}
+	}
+}
+
+// updatePosition updates the position of the given element in the sortedElements slice.
+func (s *sortedSet[ElementType, WeightType]) updatePosition(element *sortedSetElement[ElementType, WeightType]) (moved bool) {
+	// update heaviest and lightest references after we are done moving the element
+	defer func(oldIndex int) {
+		if moved {
+			if oldIndex == 0 {
+				s.heaviestElement.Set(s.sortedElements[0].element)
+			}
+
+			if oldIndex == len(s.sortedElements)-1 {
+				s.lightestElement.Set(s.sortedElements[len(s.sortedElements)-1].element)
+			}
+		}
+
+		if element.index == 0 {
+			s.heaviestElement.Set(element.element)
+		}
+
+		if element.index == len(s.sortedElements)-1 {
+			s.lightestElement.Set(element.element)
+		}
+	}(element.index)
+
+	// try to move the element to the left first
+	for ; element.index != 0; moved = true {
+		if !s.swap(s.sortedElements[element.index-1], element) {
+			break
+		}
+	}
+
+	// if the element was not moved to the left, try to move it to the right
+	if !moved {
+		for ; element.index != len(s.sortedElements)-1; moved = true {
+			if !s.swap(element, s.sortedElements[element.index+1]) {
+				break
+			}
+		}
+	}
+
+	return
+}
+
+// swap swaps the position of the two given elements in the sortedElements slice.
+func (s *sortedSet[ElementType, WeightType]) swap(left *sortedSetElement[ElementType, WeightType], right *sortedSetElement[ElementType, WeightType]) (swapped bool) {
+	if swapped = left.weight < right.weight; !swapped && left.weight == right.weight {
+		if leftAsLessable, ok := ((any)(left.element)).(lessable[ElementType]); ok {
```

Comment:
This is very implicit and not great design imo. 

As the `sortedSet` is currently implemented it allows only for `cmp.Ordered` types (see below) that implement the less method. 
Requiring `WeightType` to implement a `CompareTo` method makes this data structure more general as it allows for any object to be a type not just a derived basic type and a tie braker can be implicitly provided. Yes, it requires basic types to implement a wrapper but imo that's a sacrifice we need to make

```
type Ordered interface {
	~int | ~int8 | ~int16 | ~int32 | ~int64 |
		~uint | ~uint8 | ~uint16 | ~uint32 | ~uint64 | ~uintptr |
		~float32 | ~float64 |
		~string
}
```

Reply:
A CompareTo function will not help us as we need to hand in a variable that is monitored and triggered whenever the element weight is updated, so we will still need some "VariableType" that we can monitor.

Since the weight value also needs to be locally mirrored in the sortedSetElement, we can not treat this generically as we would somehow need to clone the entire underlying object to keep the non-updated version for our sorting.

I was experimenting with different APIs for a while and this one seemed to be the least convoluted that doesnt require us to add "Clone" methods and so on to the sorted elements.

If you have a specific implementation in mind that leads to a better API, then maybe you can propose that as a separate PR as I am not sure how to actually implement a nicer API that doesn't end up being massively convoluted.